### PR TITLE
connect: Implement NeedsLogger interface for CA providers

### PR DESCRIPTION
**Pull request by tradel**
This PR adds an interface called `NeedsLogger` to Connect CA providers. If a provider needs to log messages to the Consul log, it should implement `NeedsLogger`. Consul will then pass a preconfigured logger immediately after creating the provider, but before calling `Configure()`.

### CODE REVIEW

File: agent/consul/server_test.go [hunk 1]
```diff
@@ -115,6 +117,10 @@ func testServerConfig(t *testing.T) (string, *Config) {
 	return dir, config
 }
 
+func MakeMeATestServer(t *testing.T) (string, *Server) {
```
Comment (rboyer) [inline agent/consul/server_test.go]: This appears to not be used.

File: agent/connect/ca/provider_consul.go [hunk 2]
```diff
@@ -658,3 +660,15 @@ func (c *ConsulProvider) generateCA(privateKey string, sn uint64) (string, error
...
 	return buf.String(), nil
 }
+
+// SetLogger implements the NeedsLogger interface so the provider can log important messages.
+func (c *ConsulProvider) SetLogger(logger *log.Logger) {
+	c.logger = logger
+}
+
+// Printf sends output to the configured Logger, if one exists.
+func (c *ConsulProvider) Printf(format string, v ...interface{}) {
```
Comment (rboyer) [inline agent/connect/ca/provider_consul.go]: This appears to not be used.

File: agent/consul/server_test.go (hunk 1 above)
Comment (tradel) [reply agent/consul/server_test.go]: That's very odd. I would swear I did not write that func. Maybe it was part of someone's test code that got merged in while rebasing. I'll take it out.

File: agent/connect/ca/provider_consul.go (hunk 2 above)
Comment (tradel) [reply agent/connect/ca/provider_consul.go]: You're right, the Consul provider doesn't really do any logging currently. What happened was that I added a bunch of debugging statements while I was testing CA providers, and then removed the debug code when I was done.

File: agent/consul/server_test.go (hunk 1 above)
Comment (tradel) [reply agent/consul/server_test.go]: Fixed by 626437f553906b955871b0353930cb56aaeeeed5.

File: agent/connect/ca/provider_consul.go (hunk 2 above)
Comment (tradel) [reply agent/connect/ca/provider_consul.go]: Fixed by 7bf3e91d5a5bd15ef8e304703ac57c97cff9a53b.

**Comment by tradel** [comment]
Something went wrong when I tried to rebase off the latest master. I will close this PR and try to fix it.

**Pull request closed without merging**